persistent/Database/Persist/Class/PersistUnique.hs

```diff
@@ -249,3 +264,49 @@ checkUniqueKeys (x:xs) = do
     case y of
         Nothing -> checkUniqueKeys xs
         Just _ -> return (Just x)
+
+-- | The slow but generic 'putMany' implemetation for any 'PersistUniqueRead'.
+-- * Lookup corresponding entities (if any) for each record using 'getByValue'
+-- * For pre-existing records, issue a 'replace' for each old key and new record
+-- * For new records, issue a bulk 'insertMany_'
+defaultPutMany
+    ::( PersistEntityBackend record ~ BaseBackend backend
+      , PersistEntity record
+      , MonadIO m
+      , PersistStoreWrite backend
+      , PersistUniqueRead backend
+      )
+    => [record]
+    -> ReaderT backend m ()
+defaultPutMany []   = return ()
+defaultPutMany rsD  = do
+    let rs = nubBy ((==) `on` recordEssence) (reverse rsD)
+
+    -- lookup record(s) by their unique key
+    mEsOld <- mapM getByValue rs
+
+    -- find pre-existing entities and corresponding (incoming) records
+    let merge (Just x) y = Just (x, y)
+        merge _        _ = Nothing
+    let mEsOldAndRs = zipWith merge mEsOld rs
+    let esOldAndRs = catMaybes mEsOldAndRs
+
+    -- determine records to insert
+    let esOld = fmap fst esOldAndRs
+    let rsOld = fmap entityVal esOld
+    let rsNew = deleteFirstsBy ((==) `on` recordEssence) rs rsOld
+
+    -- determine records to update
+    let rsUpd = fmap snd esOldAndRs
+    let ksOld = fmap entityKey esOld
+    let krs   = zip ksOld rsUpd
+
+    -- insert `new` records
+    insertMany_ rsNew
+    -- replace existing records
+    mapM_ (uncurry replace) krs
+
+-- | The _essence_ of a unique record.
+-- useful for comaparing records in haskell land for uniqueness equality.
+recordEssence :: PersistEntity record => record -> [PersistValue]
+recordEssence r = concat $ map persistUniqueToValues $ persistUniqueKeys r
```

Comment:
Is there a better name for this? This seems like one of those function names where you'd always have to lookup the implementation to see what it does

Reply:
For sure, I can think of a few
- `persistUniqueKeyValues`
- `persistValuesFromAllUniqueKeys`
- `persistValuesAllUniqueKeys`

Also, this function is entirely for internal use, it is not exposed via `Database.Persist.Class`.